## Claude Review (別セッション、Opus 二段目) — 2026-05-07

**レビュアー**: Claude Opus 4.7 (1M context) / claude-opus-4-7[1m]
**対象コミット**: `453b4e3` (HEAD)
**ベース**: `c0b53ea` (origin/main)
**レビュー手順**: spec 逐条突合 + 系統的抜け漏れチェック + テスト偽陽性チェック + コード品質 + Codex 指摘対応監査
**前提**: Codex 独立レビューで MF-1 + SF-1〜4 指摘済 → fix commit `453b4e3` で対応宣言。本レビューは「対応宣言が実コードで満たされているか」+ 残留盲点の監査

---

### Must-fix

- **`backend/src/hostInfo.test.ts:39,47,55` — `npm run build` が TypeScript エラーで失敗する**
  - `vi.spyOn(fs, "readFile").mockResolvedValue("..." as unknown as Buffer)` の cast が TypeScript の overloaded signature 解決と衝突。`tsc --noEmit` / `npm run build` が以下のエラーで stop:
    ```
    src/hostInfo.test.ts(39,9): error TS2345: Argument of type 'Buffer<ArrayBufferLike>' is not assignable to parameter of type 'string | NonSharedBuffer'.
    ```
  - 検証: 本ブランチで `cd backend && npm run build` を実行 → 上記エラーで失敗。origin/main に戻すと build は通る (本 PR で新規追加したファイルが原因)。`tsconfig.json` は `src/**/*` を include し、tsc は test ファイルも対象にする。
  - vitest は `tsx`/`esbuild` で動くため**テストは pass**するが、`tsc` を経由する `npm run build` (= production 配布物 build) は壊れる。dist/ を produce する経路 (CI / package publish / docker build 等) に影響。
  - 推奨修正: `as unknown as Buffer` を **削除**して `mockResolvedValue("Linux version ...")` のみにする (`fs.readFile(path, "utf-8")` の overload は `Promise<string>` を返すので素の string で十分)。または `mockImplementation(async () => "...")` に切替。
  - 本リポは現状 `.github/workflows/` が無く CI で build を回していないが、PR template の build/test 列の sanity を破る。テスト pass のみ確認して build 検証が漏れた典型例。

### Should-fix

- **`frontend/src/components/workspace/WorkspaceListView.tsx:104-155` — race fix の **非空 path 切替時** が未対応 (Codex MF-1 の対応が部分的)**
  - 修正された経路は「path → 空文字」遷移のみ (line 137-145 の `inflightSeqRef.current++`)。しかし「path="a" の inspect inflight 中に user が "ab" まで打鍵した」遷移では、新 path への debounce setTimeout がリセットされるだけで、**進行中の `runInspect("a")` は seq guard を pass して状態を上書きしてしまう**。`runInspect` は内部で `seq=++inflightSeqRef.current` するが、**新 path に変わった時点では bump されない**ので、古い inspect の resolve が `seq===current` を満たし `setStatus(result)` 実行 → 古い path の結果が新 path の入力欄状態として表示される (UI flicker)。その 400ms 後に新 path の runInspect が走って正しい結果に収束はするが、中間状態が不整合。
  - 実装観察: 入力 "a" → 400ms → runInspect("a") 開始 (seq=1, status=inspecting) → user 即座に "ab" 入力 → useEffect 発火 → debounce timer reset (line 134-136) のみ、seq は触らず → runInspect("a") が resolve 時 seq===1=current → 古い結果で UI 更新。
  - 推奨修正: useEffect の path-change 経路 (空文字以外) でも `inflightSeqRef.current++` を実行 (= 全ての path 変更で進行中 inspect を invalidate)。あるいは `runInspect` 起動時の seq bump を debounce setTimeout 設定時に前倒し。

- **`frontend/src/components/workspace/WorkspaceListView.tsx:92` — コメント「prefix-match で suggest」が実装 (substring-match) と乖離**
  - 該当行: `// recent workspace 一覧 (store から取得、prefix-match で suggest)`
  - 実装は line 214 で `.includes(q)` = substring-match。Codex SF-2 で他箇所の用語は substring-match に統一されたが、line 92 のコメントだけ取り残されている。
  - PR 本文「ISSUE 修正案 B」節も「recent から history dropdown → `WorkspaceListView.tsx:267-290`」と書いてあるが行番号も用語も古い (実装は :286-332)。
  - 推奨修正: `// ... substring-match で suggest` に置換。

- **PR 本文「仕様逐条突合 (自己申告)」の file:line が `453b4e3` 後に shift してほぼ全件ずれている (memory `feedback_pr_self_report_line_numbers.md` 違反)**
  - 検出した主なズレ (本文 → 実コード):
    - WSL ヒント `:259-264` → 実 `:341-346`
    - debounced auto-inspect `:131-152` → 実 `:133-155`
    - 4 status badges `:316-358` → 実 `:360-421`
    - 入力欄 input `:228-241` → 実 `:246-274`
    - picker 二次ボタン `:243-252` → 実 `:275-285`
    - placeholder builder `:64-79` → 実 `:65-77`
    - filter logic `:213-220` → 実 `:209-216`
    - dropdown render `:267-290` → 実 `:286-332`
    - handleSelectRecent `:240-244` → 実 `:235-239`
    - **「ステータス=ready で『開く』ボタン」 `:373-377` → 実コード :373-377 は status="ready" badge の中身。「開く」ボタンは実 `:439-443`** (条項突合として本来「開く」を指しているはずなのに別要素を指している)
    - `hostInfo.ts:32-42` (WSL 検出) → 実 `:31-39`
  - Codex SF-1 で再計算したと PR 本文に記載があるが、その後の `453b4e3` (Escape/Tab onKeyDown 追加 + race fix の inflightSeqRef.current++ 追加 = +15 行 frontend) で全ての参照行がシフトしているにも関わらず PR 本文が更新されていない。
  - 推奨修正: PR 本文の (a)〜(c) 各条項 file:line を実コード上で再 grep して更新。

- **`frontend/e2e/workspace-os-aware-input.spec.ts:84-104` — 「recent dropdown 表示」テストが **空配列時に何も検証しない** 偽陽性パターン**
  - `if (isPresent > 0) { ... }` で recent ≥ 1 件の時だけ assertion を実行。空配列なら**何も検証せず pass** (無条件 pass)。CI の workspace 状態が空なら本テストは UI 機能を一切検証しない。
  - 推奨修正: localStorage / wsBridge mock 等で recent エントリを **必ず 1 件以上注入**してから dropdown 表示を assert する。または backend `recent-workspaces.json` を fixture で前提化。現状テスト名「dropdown が表示される」と挙動が乖離している。

### Nit / 質問

- `backend/src/hostInfo.ts:41-50` `getHostInfo()` に inflight Promise の dedup が無い。frontend 側 (`workspaceStore.ts:167-187`) は `_hostInfoInflight` で並行呼び出しを 1 つに収束しているが、backend 側は同時 2 callers が来ると `detectWSL()` を 2 回走らせる (= `/proc/version` を 2 回読む)。実害は最小だが対称性として揃えると親切。

- `backend/src/hostInfo.ts:35` の WSL 判定正規表現 `/Microsoft|WSL/i` は **docker container 内 (Linux host on Windows host)** 等で `/proc/version` が host のものを引き継ぐ場合に false positive を起こす可能性がある (たとえば Windows host 上の Docker Desktop で起動した Linux container)。本 PR の目的「WSL ユーザーに正しい placeholder 出す」は満たすが、edge case として「Docker on WSL2 内で backend を立てると isWSL=true として扱われる」は許容判定か要確認。コンテナ向けに別ヒント分岐を作るかは future work で良い。

- `frontend/src/components/workspace/WorkspaceListView.tsx:255-266` の `onKeyDown` で `e.preventDefault()` を呼んでいない。Escape / Tab を `e.key` で捕捉して `setShowRecent(false)` するが、modal 自身に Escape の global listener は無い (`tbl-modal-overlay` は onClick のみ) ので、現状は実害なし。ただし将来 modal に Escape close を追加した場合「dropdown を閉じる」と「modal を閉じる」が両方発火して closing 競合になる懸念がある。`if (showRecent) e.preventDefault();` を入れておくと将来安全。

- `frontend/src/components/workspace/WorkspaceListView.tsx:281` の picker btn は `tabIndex={-1}` で keyboard skip だが、WSL 環境では Windows file dialog が出て**そもそも機能しない**。本 PR は WSL ヒントで「フォルダ参照は使えない」と明記しているのに、ボタン自体は WSL 検出時も表示している。`{!host?.isWSL && hasPickerSupport && ...}` で WSL 時は非表示にする方が UX 一貫。

- `frontend/src/components/workspace/AddWorkspaceDialog.test.tsx` は **10 件**で pass (vitest 実測)。PR 本文「Vitest (frontend): 9 件 (新規 7 件)」と fix commit message「7 件 → 12 件」のいずれとも一致しない。文章を実数 (10 件) に揃える。

- `frontend/src/components/workspace/AddWorkspaceDialog.test.tsx:30-31` で `openWorkspaceMock` / `initAndOpenMock` を mock 登録しているが**どのテストでも assert していない**。dead mock。削除して mock factory を簡潔化できる。

- `frontend/src/components/workspace/WorkspaceListView.tsx:288` `id="workspace-recent-dropdown"` は global ID。現状 `AddWorkspaceDialog` のインスタンスは 1 つしか存在しないので衝突しないが、「同時に複数 modal が開く」将来拡張で衝突する。`useId()` を使うか、ref で参照する方が綺麗。

- `frontend/src/components/workspace/WorkspaceListView.tsx:93` `getState().workspaces` を render body 直接呼び出ししているため、ダイアログ表示中に他 tab で recent が更新されても反映されない (PR 申し送り済み)。短時間ダイアログだから許容判断は妥当。

---

### spec 逐条突合結果

#### `docs/spec/workspace.md §3.5 workspace.inspect`

| 条項 | 実装箇所 | 判定 |
|---|---|---|
| `workspace.inspect(path)` を呼び出して `ready / needsInit / notFound / invalid` 4 状態を表示 | `WorkspaceListView.tsx:106-131` (runInspect) + `:360-421` (4 status badge) + `:413-421` (error 状態) | ✓ ただし spec の 3 状態 (ready/needsInit/notFound) に invalid を実装独自で 1 つ、error を 1 つ追加で計 5 状態。spec §3.5 表の 3 状態は invalid を含まないため、スコープ外として実装が拡張している (#852 R-3 由来、同 PR で追加した訳ではない) |

#### `docs/spec/workspace.md §6.1 WebSocket リクエストメソッド`

| 条項 | 実装箇所 | 判定 |
|---|---|---|
| 既存 6 method | `wsBridge.ts:671-801` | ✓ (本 PR 改変なし) |
| **新規 `workspace.hostInfo`** を追加 | `wsBridge.ts:1373-1377` | ✓ (実装) |
| 新規 method を spec §6.1 表に追記 | spec ファイル | ❌ **spec 未更新** |

**Should-fix 候補**: `workspace.hostInfo` は本 PR で新設の WS method だが、spec `§6.1 WebSocket リクエストメソッド` 表 (line 261-267) には追記されていない。spec governance 的に schema は触らないが、spec doc は同期するべき。本 PR の rule 上は「新規 endpoint = spec doc 同期」が望ましい。  
ただし AGENTS.md の schema governance ルールは `schemas/*.json` グローバル定義に対するものであり、`docs/spec/` の追記は AI が能動的にできるはず。レビュアーとしては Should-fix に分類。

PR 本文「spec 側の不足点 / 提案: N/A」となっているが、新規 WS method 追加 = spec §6.1 追記漏れがある。

#### `docs/spec/workspace.md §6.3 MCP tools (AI エージェント用)`

| 条項 | 実装箇所 | 判定 |
|---|---|---|
| `designer__workspace_*` 6 件 (list/status/open/inspect/close/remove) | `tools.ts:1233-1284` | ✓ |
| `workspace.hostInfo` 相当の MCP tool (例: `designer__workspace_host_info`) | — | **未実装 / 議論不在** |

**質問**: `workspace.hostInfo` は **WS method のみで MCP tool にはしない** 判断か。本 PR は frontend UI 改善目的なので MCP tool 化不要は妥当だが、PR 本文に明記すると後続作業者が判断できる。Nit。

#### Codex 指摘 5 件への対応監査

| 指摘 | 対応 commit `453b4e3` の宣言 | 実コード検証 |
|---|---|---|
| MF-1: 入力空クリア時の inflight inspect seq guard 不備 | line 137-145 で `inflightSeqRef.current++` 追加 | ✓ 実装確認 (但し非空 path 切替時の race は未対応 — 上記 Should-fix 参照) |
| SF-1: PR 本文 file:line ずれ | 「Codex SF-1 で再計算済」と PR 本文記述 | ❌ **`453b4e3` のコード追加で再シフト、PR 本文未更新** (上記 Should-fix 参照) |
| SF-2: dropdown 絞り込み prefix-match 表記 vs substring-match 実装 | 用語を substring-match に統一 | △ 実装は substring-match (✓)、テスト名も substring-match (✓)、しかし `:92` の inline コメントに「prefix-match で suggest」が残存 (上記 Should-fix) |
| SF-3: dropdown が Tab/Escape で閉じない | `onKeyDown` 追加 (line 255-266) | ✓ 実装確認、test 2 件追加で検証 |
| SF-4: WSL 判定 fs mock テスト | `hostInfo.test.ts` 4 ケース追加 | ✓ vitest 7 件 pass 確認、ただし mock の Buffer cast が tsc を壊している (上記 Must-fix) |

**総括**: Codex 指摘 5 件のうち、**MF-1 は対応が部分的** (空文字遷移のみ修正、非空遷移は race window 残存)、**SF-1 は再 shift で再ずれ**、**SF-2 はコメント取り残し**、**SF-4 は対応が tsc を壊す副作用**。完全解消 ✓ は SF-3 のみ。

---

### 検証方法

- `gh pr view 919 --json` / `gh pr diff 919` で差分把握
- `git fetch origin feat/issue-858-workspace-text-input-primary` + `git log origin/main..HEAD` で 2 commits 確認 (`4124e7e` + `453b4e3`)
- `gh api repos/csilost2001/harmony/issues/919/comments` で過去レビューコメント取得 → 0 件 (Codex review は本リポ内に痕跡なし、PR 本文の宣言で確認)
- 主要ファイル読み込み: `WorkspaceListView.tsx` (全 829 行)、`AddWorkspaceDialog.test.tsx` (263 行)、`hostInfo.ts` (54 行)、`hostInfo.test.ts` (73 行)、`workspaceStore.ts` (1-220 行)、`wsBridge.ts:1360-1390` (新規 case 追加箇所)、`workspace-default-path.spec.ts` (無改変確認)、`workspace-os-aware-input.spec.ts` (新規 spec 全 104 行)、`docs/spec/workspace.md` (全 372 行)
- `cd backend && npx vitest run src/hostInfo.test.ts` → **7 件 pass** ✓
- `cd frontend && npx vitest run src/components/workspace/AddWorkspaceDialog.test.tsx` → **10 件 pass** ✓ (PR 本文記載の 9 件/12 件いずれとも乖離)
- `cd frontend && npx vitest run` → 1774 pass / 11 fail。fail は `samples-v3.schema.test.ts` 11 件で **origin/main でも同様に fail** (本 PR と無関係の pre-existing) を確認 ✓
- `cd frontend && npx eslint src/components/workspace/WorkspaceListView.tsx` → exit=0 ✓
- `cd frontend && npx tsc -p . --noEmit` → exit=0 ✓
- `cd backend && npx tsc -p . --noEmit` → **3 errors in `src/hostInfo.test.ts`** ❌ (Must-fix の根拠)
- `cd backend && npm run build` → **同上 error で失敗** ❌
- `gh pr diff 919 --name-only | grep -E "^schemas/"` → 0 件 ✓ (schema governance 違反なし)
- `git checkout origin/main && cd backend && npm run build` → 成功確認 (本 PR が build を壊した root cause を特定)
- `frontend/e2e/workspace-default-path.spec.ts` の **無改変確認** → diff には現れず ✓ (memory `feedback_no_silent_test_modification.md` 遵守)
- Playwright e2e 実行: 未実行 (UI 影響レビュー範囲だが backend 起動 + ブラウザ smoke は実施済とのこと、本セッションでは実行しない)

---

### 総合評価

**修正後マージ**

根拠:
- **Must-fix が 1 件**: `npm run build` を壊している。tsc レベルで型エラーが 3 件出るため、production 配布物 / package publish ルートに影響。修正は `as unknown as Buffer` cast の削除のみで 1〜3 行。所要 5 分。
- **Should-fix が 4 件**: race fix の非空遷移ケース不対応 (UI flicker)、コメント取り残し、PR 本文 file:line のシフト未追従、e2e 偽陽性 (空配列で何も検証しない)、spec §6.1 への新規 WS method 追記漏れ。いずれも本 PR スコープで吸収可能 (memory `feedback_pr_scope_absorb_pre_existing.md` 観点でも別 ISSUE 化は不適)。
- **Nit が 6 件**: 動作影響なし、後追い改善で良い。
- **schema governance** 違反なし ✓
- **memory `feedback_no_silent_test_modification.md`** 遵守 ✓ (`workspace-default-path.spec.ts` 無改変確認)
- **vitest pass** 確認済 (backend 7 / frontend 10 件 / 全体 origin/main と同等)

PR 本文の「仕様逐条突合 (自己申告)」が `453b4e3` で再シフトしてほぼ全件ずれている点は memory `feedback_pr_self_report_line_numbers.md` (PR #592 の繰り返し) として運用警告対象。修正後 (commit 直前) に grep で実測した値で書き直す原則を厳守。

UI 影響あり PR のため最終マージ判断はユーザー。本レビューは判断材料の提供。

--- a/backend/src/hostInfo.test.ts
+++ b/backend/src/hostInfo.test.ts
@@ -36,7 +36,7 @@ describe("getHostInfo", () => {
 
     it.runIf(isLinux)("WSL2 文字列を含む /proc/version → isWSL=true", async () => {
       vi.spyOn(fs, "readFile").mockResolvedValue(
-        "Linux version 5.15.90.1-microsoft-standard-WSL2 (oe-user@oe-host) ..." as unknown as Buffer,
+        "Linux version 5.15.90.1-microsoft-standard-WSL2 (oe-user@oe-host) ...",
       );
       const info = await getHostInfo();
       expect(info.isWSL).toBe(true);
@@ -44,7 +44,7 @@ describe("getHostInfo", () => {
 
     it.runIf(isLinux)("Microsoft 文字列 (WSL1 想定) → isWSL=true", async () => {
       vi.spyOn(fs, "readFile").mockResolvedValue(
-        "Linux version 4.4.0-19041-Microsoft (Microsoft@Microsoft.com) ..." as unknown as Buffer,
+        "Linux version 4.4.0-19041-Microsoft (Microsoft@Microsoft.com) ...",
       );
       const info = await getHostInfo();
       expect(info.isWSL).toBe(true);
@@ -52,7 +52,7 @@ describe("getHostInfo", () => {
 
     it.runIf(isLinux)("native Linux (Microsoft/WSL を含まない) → isWSL=false", async () => {
       vi.spyOn(fs, "readFile").mockResolvedValue(
-        "Linux version 5.15.0-91-generic (buildd@lcy02-amd64-009) ..." as unknown as Buffer,
+        "Linux version 5.15.0-91-generic (buildd@lcy02-amd64-009) ...",
       );
       const info = await getHostInfo();
       expect(info.isWSL).toBe(false);

--- a/docs/spec/workspace.md
+++ b/docs/spec/workspace.md
@@ -145,7 +145,26 @@ workspace.inspect(path) → { status, path, name? }
 
 **実装**: `backend/src/workspaceInit.ts:87-101`
 
-### 3.6 init=true 時の初期化内容
+### 3.6 workspace.hostInfo
+
+```
+workspace.hostInfo() → { platform, isWSL, homeDir }
+```
+
+backend が動作しているホスト OS 情報を返す。`AddWorkspaceDialog` が path 入力欄の placeholder / WSL 検出ヒント表示に使う (#858)。
+
+| field | 型 | 内容 |
+|-------|----|------|
+| `platform` | `"linux" \| "win32" \| "darwin" \| "other"` | `process.platform` を 4 値に正規化 |
+| `isWSL` | `boolean` | `process.platform === "linux"` かつ `/proc/version` に `Microsoft` または `WSL` 文字列を含む場合 `true` (大文字小文字無視) |
+| `homeDir` | `string` | `os.homedir()` (placeholder の "/home/<user>/projects/my-app" 等を組み立てるのに使用) |
+
+- backend session 内でキャッシュされる (`/proc/version` を毎回読まない)
+- `inspectWorkspacePath` のような厳密判定ではなく**情報提供 API**。frontend は描画切替の hint として使うのみで、認可判断には使わない
+
+**実装**: `backend/src/hostInfo.ts:32-49` / `backend/src/wsBridge.ts:1373-1376`
+
+### 3.7 init=true 時の初期化内容
 
 `initializeWorkspace(path)` が実行する処理:
 

--- a/frontend/src/components/workspace/WorkspaceListView.tsx
+++ b/frontend/src/components/workspace/WorkspaceListView.tsx
@@ -1,4 +1,4 @@
-import { useState, useEffect, useCallback, useMemo, useRef } from "react";
+import { useState, useEffect, useCallback, useMemo, useRef, useId } from "react";
 import { useNavigate } from "react-router-dom";
 import { mcpBridge } from "../../mcp/mcpBridge";
 import {
@@ -88,8 +88,10 @@ export function AddWorkspaceDialog({ onClose, onAdded }: AddWorkspaceDialogProps
   const debounceRef = useRef<number | null>(null);
   const inflightSeqRef = useRef(0);
   const inputRef = useRef<HTMLInputElement | null>(null);
+  // 同一画面に複数 dialog が同時表示される将来拡張に備え、global ID 衝突を避けて useId で一意化
+  const dropdownId = useId();
 
-  // recent workspace 一覧 (store から取得、prefix-match で suggest)
+  // recent workspace 一覧 (store から取得、substring-match で suggest)
   const recentWorkspaces = getState().workspaces;
 
   // host info を取得 (失敗は黙って null のまま、placeholder はフォールバックを使う)
@@ -134,10 +136,11 @@ export function AddWorkspaceDialog({ onClose, onAdded }: AddWorkspaceDialogProps
     if (debounceRef.current !== null) {
       window.clearTimeout(debounceRef.current);
     }
+    // path が変わった瞬間、進行中 inspect (旧 path 用) の遅延 response が
+    // 新 path の UI を上書きしないよう seq を bump して旧結果を破棄する。
+    // 空入力 / 非空入力切替 / 非空 → 非空切替 すべての race window をカバー。
+    inflightSeqRef.current++;
     if (!path.trim()) {
-      // 空入力に戻した瞬間、進行中 inspect の遅延 response が
-      // idle 状態を ready/notFound 等に上書きしないよう seq を bump して破棄する
-      inflightSeqRef.current++;
       setStatus("idle");
       setInspectName(null);
       setErrorMsg(null);
@@ -202,7 +205,11 @@ export function AddWorkspaceDialog({ onClose, onAdded }: AddWorkspaceDialogProps
     }
   };
 
-  const hasPickerSupport = typeof window !== "undefined" && "showDirectoryPicker" in window;
+  // WSL2 + Windows ブラウザ環境では showDirectoryPicker が Linux パスに到達できず実質使えないので、
+  // ヒント文と整合させてフォルダ参照ボタンを非表示にする (#858 / #919 Opus review nit)
+  const hasPickerSupport = typeof window !== "undefined"
+    && "showDirectoryPicker" in window
+    && !host?.isWSL;
   const placeholder = buildPlaceholder(host);
   const exampleAbs = buildOsAwareExamplePath(host);
 
@@ -222,7 +229,7 @@ export function AddWorkspaceDialog({ onClose, onAdded }: AddWorkspaceDialogProps
       const target = e.target as Node;
       if (inputRef.current && !inputRef.current.contains(target)) {
         // dropdown 内クリックは別途閉じる (handleSelectRecent)
-        const dropdown = document.getElementById("workspace-recent-dropdown");
+        const dropdown = document.getElementById(dropdownId);
         if (!dropdown || !dropdown.contains(target)) {
           setShowRecent(false);
         }
@@ -230,7 +237,7 @@ export function AddWorkspaceDialog({ onClose, onAdded }: AddWorkspaceDialogProps
     };
     document.addEventListener("mousedown", handler);
     return () => document.removeEventListener("mousedown", handler);
-  }, [showRecent]);
+  }, [showRecent, dropdownId]);
 
   const handleSelectRecent = (entry: WorkspaceEntry) => {
     setPath(entry.path);
@@ -285,7 +292,7 @@ export function AddWorkspaceDialog({ onClose, onAdded }: AddWorkspaceDialogProps
             )}
             {showRecent && filteredRecents.length > 0 && (
               <ul
-                id="workspace-recent-dropdown"
+                id={dropdownId}
                 role="listbox"
                 aria-label="最近使ったワークスペース"
                 style={{

